_posts/2018-10-12-spring-boot-angular-gradle.md

```diff
@@ -0,0 +1,132 @@
+```
+.
+├── build.gradle
+├── gradle
+│   └── wrapper
+│       ├── gradle-wrapper.jar
+│       └── gradle-wrapper.properties
+├── gradlew
+├── gradle.bat
+├── todo-api
+│   └── build.gradle
+├── todo-ui
+│   └── build.gradle
+└── settings.gradle
+```
+
+You should have three `build.gradle` files, we'll check their content within a minute. 
+But we first have to tell Gradle the name of our project and make sure it will recognize the two modules. 
+This can be done in `settings.gradle`.
+
+```groovy
+rootProject.name = 'todo'
+
+include 'todo-api', 'todo-ui'
+``` 
+
+We'll use `todo` as the project name and include both the backend and frontend module by specifying their directory name.
+It's important that this name matches the path of the directory, otherwise Gradle cannot find these modules.
+
+Gradle will now recognize both child directories as a subproject. 
+
+## Angular
+
+For the Angular part we want create a jar with a `static` directory that contains the result of our Angular build. 
+By doing this we can include the jar in our backend module. And because Spring Boot will automatically add static web 
+resources located within `static`, the Angular application will be visible when we launch the application.
+
+This can be done by using the [com.moowork.node](https://plugins.gradle.org/plugin/com.moowork.node) plugin.
+
+Let's take a look at the `build.gradle` file in our `todo-ui` project. 
+```groovy
+// 1
+plugins {
+  id 'java'
+  id "com.moowork.node" version "1.2.0"
+}
+
+// 2
+node {
+  version = '9.2.0'
+  npmVersion = '6.4.1'
+  download = true
+}
+
+// 3
+jar.dependsOn 'npm_run_build'
+
+// 4
+jar {
+  from 'dist/todo-ui' into 'static'
+}
+```
+
+Let me explain step-by-step what we're doing here:
+1. We need the `java` plugin to have the `jar` task available and the `com.moowork.node` plugin to execute node scripts like `npm_run_build`.
+2. We have to specify which node and NPM version we want to use.
+3. Before creating the jar with the `jar` task we want to build our Angular project, otherwise we don't have any static files to serve. 
+This can be done by using the `npm_run_build` task.<sup>1</sup>
+4. When we build the Angular project our static files will become available in `dist/todo-ui`. We want those files into `static`. 
+The `from 'dist/todo-ui' into 'static'` command in the `jar` task will simply copy everything from `dist/todo-ui` into `static`.<sup>2</sup>
+
+When we build the subproject it will run `npm run build` and create a new jar with the build result in the `static` directory. 
+
+We can now setup Spring Boot to include the jar. 
+
+## Spring Boot
+
+Our Spring Boot project already has a `build.gradle` file generated. 
+We just have to add one line within our dependencies to include the todo-ui module.
+
+```groovy
+dependencies {
+	implementation(project(':todo-ui'))
+}
+```
+
+Build the project, execute the generated Spring Boot jar and go to [localhost:8080](),you should see your Angular web application. 
```

Comment:
Also add `{:target="_blank" rel="noopener noreferrer"}` to the link (even though no one will probably end up clicking it :)).

Reply:
Also add a space after the second comma. 🤓